src/Processors/Formats/Impl/FreeformRowInputFormat.cpp

```diff
@@ -0,0 +1,473 @@
+{
+    if (in.eof())
+        return {};
+
+    if (*in.position() != '{')
+    {
+        String ret;
+        readJSONField(ret, in);
+        return {{fmt::format("c{}", index), ret}};
+    }
+
+
+    ++in.position();
+    skipWhitespacesAndDelimiters(in);
+
+    std::vector<std::pair<String, String>> ret;
+    while (*in.position() != '}')
+    {
+        String field_name = JSONUtils::readFieldName(in);
+
+        auto fields = parseFields(in, index); // "index" here doesn't do anything
+
+        for (const auto & [name, field] : fields)
+        {
+            if (name.starts_with("c"))
+                ret.emplace_back(field_name, field);
+            else
+                ret.emplace_back(fmt::format("{}.{}", field_name, name), field);
+        }
+
+        skipWhitespacesAndDelimiters(in);
+    }
+
+    ++in.position();
+
+    return ret;
+}
+
+std::vector<std::pair<String, String>> CSVFieldMatcher::parseFields(ReadBuffer & in, size_t index) const
+{
+    String ret;
+    readCSVField(ret, in, settings.csv);
+    return {{fmt::format("c{}", index), ret}};
+}
+
+std::vector<std::pair<String, String>> QuotedFieldMatcher::parseFields(ReadBuffer & in, size_t index) const
+{
+    String ret;
+    readQuotedField(ret, in);
+    return {{fmt::format("c{}", index), ret}};
+}
+
+std::vector<std::pair<String, String>> EscapedFieldMatcher::parseFields(ReadBuffer & in, size_t index) const
+{
+    String ret;
+    readEscapedString(ret, in);
+    return {{fmt::format("c{}", index), ret}};
+}
+
+std::vector<std::pair<String, String>> RawByWhitespaceFieldMatcher::parseFields(ReadBuffer & in, size_t index) const
+{
+    String ret;
+    readStringUntilWhitespaceDelimiter(ret, in);
+    return {{fmt::format("c{}", index), ret}};
+}
+
+FreeformFieldMatcher::FreeformFieldMatcher(ReadBuffer & in_, const FormatSettings & settings_)
+    : max_rows_to_check(std::min<size_t>(100, settings_.max_rows_to_read_for_schema_inference)), in(in_)
+{
+    // matchers are pushed in the order of priority, this helps with exiting early and reducing the search tree.
+    matchers.emplace_back(std::make_unique<JSONFieldMatcher>(FormatSettings::EscapingRule::JSON, settings_));
+    matchers.emplace_back(std::make_unique<CSVFieldMatcher>(FormatSettings::EscapingRule::CSV, settings_));
+    matchers.emplace_back(std::make_unique<RawByWhitespaceFieldMatcher>(FormatSettings::EscapingRule::Raw, settings_));
+    matchers.emplace_back(std::make_unique<QuotedFieldMatcher>(FormatSettings::EscapingRule::Quoted, settings_));
+    matchers.emplace_back(std::make_unique<EscapedFieldMatcher>(FormatSettings::EscapingRule::Escaped, settings_));
+}
+
+std::vector<FreeformFieldMatcher::Fields> FreeformFieldMatcher::readNextFields(bool one_string, size_t index)
+{
+    skipWhitespacesAndDelimiters(in);
+    char * start = in.position();
+    std::vector<FreeformFieldMatcher::Fields> next_fields;
+
+    if (one_string)
```

Comment:
Does this flag mean that we should treat all data up to `\n` as one `String` field?
Maybe not a good name for the flag.